### PR TITLE
Add API and Webhook for PodMarker

### LINKED ISSUE #640: [feature request] PodMarker: mark Pods by number, labels, nodes, probes
**What would you like to be added**:

PodMarker: A new CRD that marks Pods by number, labels, nodes, probes.

**Why is this needed**:

- A canary release plan may want to upgrade X number of Pods out of Y number of total replica.
- Differentiate Pods with labels according to their different Nodes.
- Mark Pods that satisfy the specific prober (http/tcp/exec).

### PR DESCRIPTION
**Pull request by veophi**
Signed-off-by: veophi <vec.g.sun@gmail.com>

### Ⅰ. Describe what this PR does
Add api and webhooks for podmarker.

**The codes were tested locally, but `make manager` and `make manifests` are not executed in this PR for better review.**

### Ⅱ. Does this pull request fix one issue?

fixed #640

### CODE REVIEW

File: pkg/webhook/podmarker/validating/podmarker_create_update_handler.go [hunk 1]
```diff
@@ -0,0 +1,188 @@
...
+	return append(allErrs, validatePodMarkerMarkIdentities(&spec.MarkItems, fldPath.Child("markIdentities"))...)
+}
+
+func validatePodMarkerStrategy(strategy *appsv1alpha1.PodMarkerStrategy, fldPath *field.Path) (allErrs field.ErrorList) {
+	// 1.ConflictPolicy must be "", "Ignore", or "Overwrite"
+	switch strategy.ConflictPolicy {
+	case "", appsv1alpha1.PodMarkerConflictOverwrite, appsv1alpha1.PodMarkerConflictIgnore:
+	default:
+		allErrs = append(allErrs, field.Invalid(fldPath.Child("conflictPolicy"), strategy.ConflictPolicy,
+			fmt.Sprintf("unsportted ConflictPolicy type: %s, ConflictPolicy should be Ignore or Overwrite", strategy.ConflictPolicy)))
```
Comment (furykerry) [inline pkg/webhook/podmarker/validating/podmarker_create_update_handler.go:83]: unsportted  -> unsupported

File: pkg/webhook/podmarker/validating/podmarker_create_update_handler.go [hunk 2]
```diff
@@ -0,0 +1,188 @@
...
+	return
+}
+
+func validatePodMarkerMarkIdentities(markItems *appsv1alpha1.PodMarkerMarkItems, fldPath *field.Path) (allErrs field.ErrorList) {
+	// markItems cannot be empty
+	if len(markItems.Labels)+len(markItems.Annotations) == 0 {
+		return append(allErrs, field.Invalid(fldPath, *markItems, fmt.Sprintf("markItems cannot be empty")))
+	}
+	// 1. validate the labels users want to mark
+	allErrs = metav1validation.ValidateLabels(markItems.Labels, fldPath.Child("labels"))
```
Comment (furykerry) [inline pkg/webhook/podmarker/validating/podmarker_create_update_handler.go:141]: is possible the marked labels will make match selector invalid?

Comment (veophi) [reply pkg/webhook/podmarker/validating/podmarker_create_update_handler.go:141]: sounds very reasonable

File: pkg/webhook/podmarker/validating/podmarker_create_update_handler.go (hunk 1 above)
Comment (veophi) [reply pkg/webhook/podmarker/validating/podmarker_create_update_handler.go]: done

**Review by furykerry — COMMENTED** [review]
/lgtm

File: pkg/util/selector.go [hunk 3]
```diff
@@ -30,8 +30,8 @@ import (
 //   * a=b,c=d,e=f and a=x,c=d,e=f
 // then others is overlap：
 //   * a=b and c=d
-func IsSelectorOverlapping(selector1, selector2 *metav1.LabelSelector) bool {
-	return !(isDisjoint(selector1, selector2) || isDisjoint(selector2, selector1))
+func IsDefinitelyNeverOverlapped(selector1, selector2 *metav1.LabelSelector) bool {
```
[the remark this answers is not in the document]
Comment (veophi) [reply pkg/util/selector.go:33]: `IsConflicting`

File: pkg/webhook/podmarker/validating/podmarker_create_update_handler_test.go [hunk 4]
```diff
@@ -0,0 +1,316 @@
...
+	decoder, _ := admission.NewDecoder(testScheme)
+	markerHandler := &PodMarkerCreateUpdateHandler{Decoder: decoder}
+
+	cases := []struct {
+		name                 string
+		getPodMarker         func() *appsv1alpha1.PodMarker
+		isExpectedSuccessful func() bool
+	}{
+		{
+			name: `normal case, in case of matchPodSelector={"a"="b"}, markItems={"c"="d"}`,
```
[the remark this answers is not in the document]
Comment (veophi) [reply pkg/webhook/podmarker/validating/podmarker_create_update_handler_test.go:93]: done

**Review by furykerry — COMMENTED** [review]
/lgtm

**Pull request closed without merging**